dask_expr/_merge.py

```diff
@@ -37,6 +37,7 @@ class Merge(Expr):
         "suffixes",
         "indicator",
         "shuffle_backend",
+        "_meta",
```

Comment:
> We genuinely change the object when we re-create it, which means that we will always trigger a fresh computation of meta

If the inputs to the pd.merge call are going to change then it seems like we need to recompute meta anyway.  If the inputs aren't changing but we're recomputing then maybe that is a sign that we're caching on the wrong things.  Maybe we should have a `staticfunction` or something instead.  

> We can simply not pass meta in this case which would trigger a fresh computation, this is only a fast path

Imagine an optimization which did something like the following:

```python
def subs(obj, old, new):
    operands = [
        new if operand == old else operand  # substitute old for new in all operands
        for operand in obj.operands
    ]
    return type(obj)(*operands)
```

This fails if we store derived state in operands, because `_meta` is in there and really we should choose to not include it any more. 

I'm ok keeping derived state on the class.  I just don't think that it should be in operands.  This probably requires custom constructors.

Reply:
There are optimisations where meta changes but we compute the necessary information anyway, so we can simply adjust meta if we want to, Projections are a good example for this. 

Lower is an example where meta won't change in case of merge, but you can't properly cache it either. We might introduce a shuffle which means caching will fail for most cases.

I'll add a custom constructor here that should work as well. Then we can see how that looks

Reply:
Sorry, my comment wasn't clear enough probably. That's how I understood what you were saying, but the problem is as follows:

- That won't work if we introduce a Projection (which is something we could accept, although that wouldn't make me happy in this particular case)
- Meta won't change when we lower the graph, but we will introduce a shuffle, so the inputs of the merge computation will change while lowering, which means that the cache wouldn't work anymore

Reply:
Feels like we're maybe talking past each other here.  Happy to chat live if you're free later.